Add settings language, gpt model, get a detailed answer as html

KBobovskiy opened the pull request:
Hello, John,

Thank you for creating and publishing this extension.

I want to know if you're interested in these changes, or should I keep them to myself?

Implemented features in the extension:

Added a setting for the GPT model (default value: GPT-4o).
Added display of the model name after translation.
Added a setting for the target language for translation (default value: English).
Added an option to provide detailed explanations with examples for phrases (with a "rough" check for no more than four words or 50 characters), with output in HTML format.
Added styles for the settings pop-up window.
Fixed the HTML layout.
Added a language code hint from the page data to the GPT instruction.

TODO:
Update _README.md_.
Create a setting for the _max_tokens_ value.

Commit:
Add extension settings

Language, GPT Model, Get a detailed answer as HTML.

Commit:
Fix quotes, remove commented code.

Commit:
Fix pop-up styles

johnlewissims commented:
@KBobovskiy Thank you for contributing to and improving this project! I really appreciate your additions, particularly:

- Support for different models
- Multi-language capabilities
- Addition of examples

I've identified a few items to address before merging, mostly cosmetic in nature.

**System Instructions Issue**
There seems to be some confusion in how the model interprets its instructions. Here are two examples of unexpected behavior when requesting translations:
<img width="975" alt="Screenshot 2025-01-04 at 1 05 59 PM" src="https://github.com/user-attachments/assets/70d05b04-e2e3-4a24-b52f-a52011e41f27" />
Another example in English...
<img width="979" alt="Screenshot 2025-01-04 at 1 06 56 PM" src="https://github.com/user-attachments/assets/09c2a01b-1c69-4249-a1da-72856918a0df" />

Here are my settings...
<img width="878" alt="Screenshot 2025-01-04 at 1 10 27 PM" src="https://github.com/user-attachments/assets/7f8acd5e-c18a-495f-a013-3a9746207da9" />

**Model Selection Enhancement**

Having the ability to change models is a huge help. I think it might make sense though to list the [available option](https://api.openai.com/v1/models) via the API. This would be easier and remove the possibility of someone entering a typo. 

The user would need to add their API token first, so we might want to make this field conditional upon the value of API token field not being null.

**Changing the Language Selector to a Dropdown** 
This is a feature that I had considered at first, the issue that I was worried about were cases where a language wasn't well supported by the model. 

What would be nice would be to add a dropdown with a list of languages that we are sure will work (maybe the top 50 most spoken). We could then provide those as options.

**Final Thoughts**

Out of these three things, only the first one seems like a blocker to me. The last two are thing that could be addressed at a later point if this is something you'd rather not worry about right now. Please let me know your thoughts.

Overall, very excited about your contribution and am eager to merge these in!

johnlewissims reviewed (changes requested):
See comment above ☝️

Commit:
Add dropdowns for Models and Language. Split detailed answer and HTML answer. Interface improvements.

Commit:
Remove debug code

Commit:
Add searching language code from the nearest parent of the selected text.

KBobovskiy commented:
@johnlewissims Hi, I've fixed your issues and added some improvements,

**

- [x] **System Instructions Issue**.
I separated the settings for the detailed answer and HTML formatting and disabled the HTML format for queries with the max_tokens value of less than 300. 
![image](https://github.com/user-attachments/assets/c731fa32-61e4-4895-8558-fe15823b6bb3)

GPT returns different results for one query. The testing query is: わかります。 from the [page](https://www.transparent.com/learn-japanese/phrases). It's the last phrase on the page.

1. Max_tokens=100. I've got a simple answer:
![image](https://github.com/user-attachments/assets/86f48db1-cedf-4adf-9081-a91843b32ce4)

2. Max_token=500. I've got a more complicated answer but there's no Explanation.
![image](https://github.com/user-attachments/assets/abc8a3fd-3b76-48fb-913e-45dddf66128e)

3. Max_token=500. As the second this answer also doesn't contain the Explanation.
![image](https://github.com/user-attachments/assets/29063df2-e0b8-448a-a99a-44f3adc79c59)

4. Max_token=500. The answer contains the Explanation.
![image](https://github.com/user-attachments/assets/f03a5f63-88cf-4eaf-b5e1-827c6d561115)

Added functionality to search for the "lang" attribute in the nearest parent of the selected text element.  But that doesn't solve the problem completely. Because Google marks its search results as English, although it can contain different languages:
![image](https://github.com/user-attachments/assets/cfc1bc46-c346-4f16-baf6-c8ff2ff17aaf)
Google marked this block with **lang=en**
![image](https://github.com/user-attachments/assets/4a5aa005-01ac-43b6-98e0-8bbd47cbcce0)
So in cases where the language hint matches the target language, the language hint is ignored.

**In summary**, it would be good to find clearer instructions for GPT to get more predictable answers to the same query.

- [x] **Model Selection Enhancement**. But I'm not sure they all should be on the list.
![image](https://github.com/user-attachments/assets/29e0b321-8d72-465e-b1b7-72064b49237b)
- [x] **Changing the Language Selector** to a Dropdown. Add the dropdown with languages from ISO 639-1. Languages listed on this [page](https://www.w3schools.com/tags/ref_language_codes.asp)

- [x] Added setting for **Maximum tokens**

![image](https://github.com/user-attachments/assets/dd871d09-034e-4255-8004-2a7af349bf3f)

- [x] Added the information about spent tokens after the model name. Added original text into the pop-up header:

![image](https://github.com/user-attachments/assets/3a597c09-f1c7-419d-ad0f-ed23db23e3f7)

Commit:
bugfix. compare language from hint and target language

johnlewissims reviewed (changes requested):
@KBobovskiy Much better! We're very close now. Just a couple more small things I wanted to double check with you on before publishing.

**1) Do we need to display the technical details after each translation?** 
<img width="899" alt="Screenshot 2025-01-06 at 2 38 58 PM" src="https://github.com/user-attachments/assets/47e3aba7-e9c2-427c-a360-4a281838b2de" />
"gpt-4o, tokens prompt:47, completion:14, total:61" 
Seems like something we could enable for a user based on a setting maybe? I certainly like know this information, but I don't think most of the users will need it. 

If we do decide to keep it, we should probably remove it from either the translation or the explanation instead of keeping it in both. 

<img width="702" alt="Screenshot 2025-01-06 at 2 42 54 PM" src="https://github.com/user-attachments/assets/9f369a33-892f-4416-8a58-567e6ada6dff" />

**2) HTML Answer Setting**

I think it's OK to set the default here to true (and the Maximum tokens setting to 300). Most users will want the improved quality of the response and the cost for most models is extremely affordable. I'm happy to defer to you on this point though if you / the community feel differently.

**3) What is the purpose of getDetailedTranslation?**

I've enabled and disabled it several times, but I don't see any real difference in the response. It looks like from the code that it is used to determine if the response is supposed to be full HTML or not. I would advise either changing the prompt based on this setting or removing it all together. 

Please let me know though if I'm miss-understanding.

**4) Add yourself to the readme as a contributor**

If you'd like of course. Would love for you to get credit for this.

Thank you again for the hard work here!

Commit:
Fix instructions for GPT, Language hind, Styles.

KBobovskiy commented:
@johnlewissims, here are my thoughts about your suggestions:
**1) Do we need to display the technical details after each translation?**
I agree with you. It would be better to hide this information under a new setting flag.

**3) What is the purpose of getDetailedTranslation?**
I think we should replace this flag with a second icon or button next to ![image](https://github.com/user-attachments/assets/d04644b1-4783-4a93-b25e-555726a03326) as it's not clear when this flag takes effect on the translation. The idea was that long phrases or sentences should be translated as a text:
![image](https://github.com/user-attachments/assets/b8a88852-7870-4d5c-9670-cef3782d9535)

But a short phrase or a single word should be translated by providing additional information such as "usage options" and "examples":
![image](https://github.com/user-attachments/assets/0267308b-1c5b-495d-8e1a-dad3ec0fb8ae)

Moreover, GPT generates irrelevant examples for a sentence:
![image](https://github.com/user-attachments/assets/d2bf5af5-1b85-4d80-815e-2d773c7e8e6e)

Unfortunately, I have no idea how to check that the selected text is a short phrase. Especially for Chinese and other languages ​​with hieroglyphs. The current check is too simple and overly fragile:
`canRequestDetailDescription(text) { return text && text?.length < 50 && text.split(' ').length < 5; }`
It would be better if users could choose what type of translation they want.

**Instructions for GPT** I've updated a set of instructions. These changes allow us to get more predictable results with some "boilerplate" in an answer (like: "1: Detected language", "2: Translation to English" etc).
```
    getOutBlocksMessages(language, isDetailedTranslation = false) {
        if (isDetailedTranslation) {
            return [
                {role: "user", content: "Output must contain 4 blocks."},
                {role: "user", content: "1: Detected language." },
                {role: "user", content: `2: Translation to ${language}.` },
                {role: "user", content: `3: Description of the options for using the text in ${language}.` },
                {role: "user", content: `4: Three examples with text in detected language with translation to ${language}.` },
                {role: "user", content: `Block headers must be in ${language}}.` }
            ];
        }

        return [
            {role: "user", content: "Output must contain 2 blocks."},
            {role: "user", content: "1: Detected language." },
            {role: "user", content: `2: Translation to ${language}}.` },
            {role: "user", content: `Block headers must be in ${language}}.` }
        ];
    }
```

I pushed my changes with new instructions for GPT.

Commit:
Add setting: displayTokens

KBobovskiy commented:
@johnlewissims Hi. I've added a new setting: **Display GPT Model and used tokens** with default value false.

Commit:
Replace the getDetailedTranslation setting with an additional icon.

KBobovskiy commented:
@johnlewissims I've removed the `getDetailedTranslation` setting and added a new icon next to the first one to request a detailed translation.
![image](https://github.com/user-attachments/assets/dc13217e-84e3-4715-8245-e7fff922e2c8)

Commit:
Hide a detailed translation icon if the selected text is to long

Commit:
Add a check to verify that the extension's context has been destroyed

johnlewissims commented:
@KBobovskiy Great work here! I think we're getting very close. 

I love the idea of a short and long translation. In Chinese (probably all pictographic languages), what readers really need is the ability to know where and how a _single character_ is used. Phrases have a similar use case, but by far what I find my self struggling with is remembering where I've seen a character before.

The only thing I see here that might need to be addressed before merge...

**Can we get the "See Explanation" link removed when looking at a detailed explanation?**

<img width="728" alt="Screenshot 2025-01-09 at 5 31 20 PM" src="https://github.com/user-attachments/assets/c02c246e-f701-4560-a147-02512c6a0913" />

"See Explanation" really isn't needed in cases where we've already given the user so much detail.

Commit:
Remove the explanation link from the detailed translation pop-up. Small code improvements.

KBobovskiy commented:
@johnlewissims  I've removed "See Explanation" link.

I have another idea: add a link to the extension settings on the pop-up form with translation. This link will only be displayed if the user has any extension settings that are not assigned a value. This way, we can automatically and visually notify users that new settings have been added. So after the user opens the settings and clicks the 'Save' button, all settings will be updated, and the link will no longer be displayed.

![image](https://github.com/user-attachments/assets/e86e52c8-ae6e-46ad-8a23-2f76bdb0b672)

johnlewissims reviewed (approved):
Thanks @KBobovskiy ! Great work here.
